File: backend/src/test/java/be/lutske/leolegacy/interfaceadapter/rest/RecipeResourceTest.java
package be.lutske.leolegacy.interfaceadapter.rest;

import io.quarkus.test.junit.QuarkusTest;
import org.junit.jupiter.api.Test;

import static io.restassured.RestAssured.given;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.Matchers.greaterThanOrEqualTo;
import static org.hamcrest.Matchers.notNullValue;

/**
 * Integration test for the /api/recipes endpoint.
 *
 * Verifies list, detail, filter-by-category, and top recipes.
 */
@QuarkusTest
class RecipeResourceTest {

    @Test
    void getApiRecipesReturns200WithJsonList() {
        given()
                .when()
                .get("/api/recipes")
                .then()
                .statusCode(200)
                .contentType("application/json")
                .body("size()", greaterThanOrEqualTo(1));
    }

    @Test
    void getApiRecipesReturnsRecipesWithExpectedFields() {
        given()
                .when()
                .get("/api/recipes")
                .then()
                .statusCode(200)
                .body("[0].id", is(greaterThanOrEqualTo(1)))
                .body("[0].title", is(notNullValue()))
                .body("[0].categoryName", is(notNullValue()));
    }

    @Test
    void getApiRecipesWithCategoryIdFilterReturnsFilteredResults() {
        given()
                .queryParam("categoryId", 1)
                .when()
                .get("/api/recipes")
                .then()
                .statusCode(200)
                .contentType("application/json")
                .body("size()", greaterThanOrEqualTo(1));
    }

    @Test
    void getApiRecipesTopReturnsTopRecipes() {
        given()
                .when()
                .get("/api/recipes/top")
                .then()
                .statusCode(200)
                .contentType("application/json");
    }

    @Test
    void getApiRecipesIdReturnsRecipeDetail() {
        given()
                .when()
                .get("/api/recipes/1")
                .then()
                .statusCode(200)
                .contentType("application/json")
                .body("id", is(1))
                .body("title", is(notNullValue()))
                .body("ingredients.size()", greaterThanOrEqualTo(1))
                .body("preparation", is(notNullValue()))
                .body("categoryName", is(notNullValue()));
    }

    @Test
    void getApiRecipesIdReturns404ForUnknownId() {
        given()
                .when()
                .get("/api/recipes/999999")
                .then()
                .statusCode(404);
    }
}

File: backend/src/test/kotlin/be/lutske/leolegacy/interfaceadapter/rest/RecipeResourceTest.kt
package be.lutske.leolegacy.interfaceadapter.rest

import io.quarkus.test.junit.QuarkusTest
import io.restassured.RestAssured.given
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.Matchers.greaterThanOrEqualTo
import org.hamcrest.Matchers.notNullValue
import org.junit.jupiter.api.Test

/**
 * Integration test for the /api/recipes endpoint.
 *
 * Verifies list, detail, filter-by-category, and top recipes.
 */
@QuarkusTest
class RecipeResourceTest {

    @Test
    fun `GET api-recipes returns 200 with JSON list`() {
        given()
            .`when`()
            .get("/api/recipes")
            .then()
            .statusCode(200)
            .contentType("application/json")
            .body("size()", greaterThanOrEqualTo(1))
    }

    @Test
    fun `GET api-recipes returns recipes with expected fields`() {
        given()
            .`when`()
            .get("/api/recipes")
            .then()
            .statusCode(200)
            .body("[0].id", `is`(greaterThanOrEqualTo(1)))
            .body("[0].title", `is`(notNullValue()))
            .body("[0].categoryName", `is`(notNullValue()))
    }

    @Test
    fun `GET api-recipes with categoryId filter returns filtered results`() {
        given()
            .queryParam("categoryId", 1)
            .`when`()
            .get("/api/recipes")
            .then()
            .statusCode(200)
            .contentType("application/json")
            .body("size()", greaterThanOrEqualTo(1))
    }

    @Test
    fun `GET api-recipes-top returns top 10 recipes`() {
        given()
            .`when`()
            .get("/api/recipes/top")
            .then()
            .statusCode(200)
            .contentType("application/json")
    }

    @Test
    fun `GET api-recipes-id returns recipe detail`() {
        given()
            .`when`()
            .get("/api/recipes/1")
            .then()
            .statusCode(200)
            .contentType("application/json")
            .body("id", `is`(1))
            .body("title", `is`(notNullValue()))
            .body("ingredients.size()", greaterThanOrEqualTo(1))
            .body("preparation", `is`(notNullValue()))
            .body("categoryName", `is`(notNullValue()))
    }

    @Test
    fun `GET api-recipes-id returns 404 for unknown id`() {
        given()
            .`when`()
            .get("/api/recipes/999999")
            .then()
            .statusCode(404)
    }
}
